packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+        );
+
+        // Store and log proposed distribution.
+        proposals[proposalIndex] = Proposal({
+            rewardIndex: rewardIndex,
+            timestamp: timestamp,
+            merkleRoot: merkleRoot,
+            ipfsHash: ipfsHash
+        });
+        emit ProposalCreated(
+            reward.sponsor,
+            reward.rewardToken,
+            rewardIndex,
+            proposalIndex,
+            timestamp,
+            reward.maximumRewardAmount,
+            merkleRoot,
+            ipfsHash
+        );
+
+        // Bump nextCreatedProposal index.
+        nextCreatedProposal++;
+    }
+
+    /**
+     * @notice Allows any caller to execute distribution that has been validated by the Optimistic Oracle.
+     * @param proposalIndex Index for identifying existing rewards distribution proposal.
+     * @dev Calling this for unresolved proposals will revert. Both accepted and rejected distribution
+     * proposals will be deleted from storage.
+     */
+    function executeDistribution(uint256 proposalIndex) external nonReentrant() {
+        // All valid proposals should have non-zero proposal timestamp.
+        Proposal memory proposal = proposals[proposalIndex];
+        require(proposal.timestamp != 0, "Invalid proposalIndex");
+
+        // Append proposal index to ancillary data.
+        Reward memory reward = rewards[proposal.rewardIndex];
+        bytes memory ancillaryData = _appendProposalIndex(proposalIndex, reward.customAncillaryData);
+
+        // Get resolved price. Reverts if the request is not settled or settleable.
+        int256 resolvedPrice =
+            optimisticOracle.settleAndGetPrice(reward.priceIdentifier, proposal.timestamp, ancillaryData);
+
+        // Transfer rewards to MerkleDistributor for accepted proposal.
+        if (resolvedPrice == 1e18) {
+            reward.rewardToken.safeApprove(address(merkleDistributor), reward.maximumRewardAmount);
+            merkleDistributor.setWindow(
+                reward.maximumRewardAmount,
+                address(reward.rewardToken),
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+            emit RewardDistributed(
+                reward.sponsor,
+                reward.rewardToken,
+                proposal.rewardIndex,
+                proposalIndex,
+                reward.maximumRewardAmount,
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+        } else emit ProposalRejected(proposal.rewardIndex, proposalIndex);
+
+        // Delete resolved proposal from storage. This also avoids double-spend for approved proposals.
+        delete proposals[proposalIndex];
+        emit ProposalDeleted(proposal.rewardIndex, proposalIndex);
+    }
+
+    /**
+     * @notice Allows any caller to delete distribution that was rejected by the Optimistic Oracle.
+     */
+    function deleteRejectedDistribution(uint256 proposalIndex) external nonReentrant() {}
```

Comment:
This is not implemented yet, but since rejected proposals get deleted in `executeDistribution`, not sure if separate call is required, except for allowing to reduce state bloat. But who would be motivated to call that anyways?

Reply:
I think the enum makes more sence but we can wait for feedback from others on this.